Sketch signature-based retry configs (#9)
grittygrease/draft-sullivan-tls-implicit-ech: draft-sullivan-tls-implicit-ech.md

The extension has the following format:

~~~~
    enum { implicit_ech_config (TBD), (65535) } ECHConfigExtensionType;

    struct {

>       HashAlgorithms hash_algorithm_id; // Needs a new IANA registry

Reviewer: ```suggestion
      HashAlgorithm hash_algorithm_id; // Needs a new IANA registry
```

The number of hash algorithm registries that IANA maintains is potentially worth another XKCD 927-style joke.

Thankfully, we have https://www.iana.org/assignments/tls-parameters/tls-parameters.xhtml#tls-parameters-18 which can be revived.

>       SignatureScheme signature_algorithm_id; // As in TLS defined in RFC 8446 4.2.3

Reviewer: ```suggestion
      SignatureScheme signature_algorithm_id; // Defined in RFC 8446 4.2.3
```

>       opaque retry_signing_key_hash<1..2^16-1>;

Reviewer: ```suggestion
      opaque retry_signing_key_hash<1..255>;
```

    } RetryKey;

    struct {

>       RetryKey retry_keys<1..255>

Reviewer: ```suggestion
      RetryKey retry_keys<1..255>;
```

    } ImplicitECHConfig;
~~~~

The presence of this extension in the
ECHConfig signals to the client that the client-facing server is configured
for implicit ECH and follows the requirements of this document.

The extension conveys a list of public key hashes which correspond to public keys which can
be used to authenticate ECH retry configs.

## Implicit ECH Authenticator Extension

A new ECHConfig extension type is defined to authenticate implicit mode retry configs.

~~~~
    enum { implicit_ech_auth (TBD), (65535) } ECHConfigExtensionType;

    struct {
      uint32 not_after_timestamp; // Linux Epoch timestamp

>       SignatureScheme signature_algorithm_id; // As in TLS defined in RFC 8446 4.2.3

Reviewer: ```suggestion
      SignatureScheme signature_algorithm_id; // Defined in RFC 8446 4.2.3
```

>       HashAlgorithms hash_algorithm_id; // Needs a new IANA registry

Reviewer: ```suggestion
      HashAlgorithm hash_algorithm_id; // Needs a new IANA registry
```

      opaque public_key<1..2^16-1>; // Parsed via signature_algorithm_id
      opaque signature<1..2^16-1>; // Parsed via signature_algorithm_id
    } ImplicitECHAuthenticator;
~~~~

Clients use the signature conveyed in this extension to authenticate the provided retry configs in the EncryptedExtension message.

The signature is computed over the message:

~~~~
  "ECH_implicit_retry_config_authenticator" || ech_config_payload
~~~~

The `ech_config_payload` consists of the ECHConfig itself, including all fields except the Signature part of the ImplicitECHAuthenticator. The ImplicitECHAuthenticator should always appear as the last extension (similar to the PSK extension in TLS).

TODO: There's a choice between having the ImplicitECHAuthenticator be an ECHConfig extension, conveyed inside retry_configs or
its own TLS extension. The latter is a bit cleaner, but violates the request-response pattern of TLS extensions unless we also
bump to a new ECH version. This also complicates how the server tells the client that ECH should be disabled. Putting the authenticator

> inside the config means that we also need to define a variant of the config with an empty body to indicate "ECH should be disabled."

Reviewer: Could this be a new message type, like NewSessionTicket, with the ECH extension only signaling that a new config will be provided?

still picks a cipher suite from key_config.cipher_suites, produces a valid HPKE
ephemeral key, and encrypts ClientHelloInner into the payload field.

> If ECH is accepted, then the same procedure as-in draft-ietf-tls-esni is followed.

Reviewer: ```suggestion
If ECH is accepted, then the same procedure as-in {{I-D.ietf-tls-esni}} is followed.
```

However, is ECH is rejected and the Implict ECH was advertised, the client must follow
a different strategy detailed in the next section.

## Behavior when ECH fails

Firstly, the client looks at the server's EncryptedExtensions message and checks for the presence
of an ECH extension with retry_configs. If present, the client extracts any provided ECHConfigs
which have the ImplicitECHAuthenticator extension and ignores any without it.

> If the extension is present, the client must hash the provided PublicKey using indicated hash algorithm and check that it appears in the HashedPublicKey list provided in the initialImplicitECH extension. If so,

Reviewer: It is not 100% clear that this refers to the extension that the client received via other means (either a DNS SVCB lookup or a previous retry config.  There's a potential for this to be mistaken to refer to the value in the retry config, which would not work.

it must then validate the provided signature against that public key.

If the signature verifies, the client treats the retry_configs as authentic and uses them to retry

> as specified in draft-ietf-tls-esni Section 6.1.6 without considering the server's TLS certificate.

Reviewer: ```suggestion
as specified in {{I-D.ietf-tls-esni, Section 6.1.6}} without considering the server's TLS certificate.
```

keys supported by a server on the same IP address since the outer SNI and
config_id can no longer be used to choose the appropriate ECH configuration.

Deployments can support key rotation using the following strategy:

>  * Add a new hashed public key to their ImplicitECHConfig

Reviewer: It looks like the procedures you define won't allow this to be propagated without causing ECH failures.  At least in the transition from public_name to key hashes.

It wasn't obvious to me that the text on Enterprise policy was intended to encompass the existing public_name-based retry method.  If that is the case, then this probably needs to be clearer.

Author: It's not explicit in the draft, but I think servers can support traditional ECH and implicit ECH simultaneously. Clients that don't know about Implicit will ignore the extension (and work fine).

 * Wait for the new ECHConfig to propagate (DNS TTL)
 * Begin signing their retry_configs with the new public key.
 * Remove the old public key hash from their ImplicitECHConfig.

Deployments can handle key compromise by removing the affected public key
from their ImplicitECHConfig.

> notAfter times SHOULD be at least 24 hours in the future and no more than 48 hours in the future. This allows

Reviewer: I'd extend this a fair bit longer than 48 hours.  The value of the thing being protected here is far less than what a certificate protects and those last for longer.

in use on the same server. Operators should consider minimizing the number
of active keys to mitigate this cost.

## Authentication of Retry Configs

> draft-ietf-tls-esni specifies how retry_configs are authenticated based on the public_name specified

Reviewer: ```suggestion
{{I-D.ietf-tls-esni}} specifies how retry_configs are authenticated based on the public_name specified
```
